oauth2_http/java/com/google/auth/oauth2/Slf4jLoggingHelpers.java

```diff
@@ -0,0 +1,180 @@
+
+/** Contains helper methods to log auth requests and responses */
+class Slf4jLoggingHelpers {
+  private static final Gson gson = new Gson();
+  private static final Set<String> SENSITIVE_KEYS = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+
+  static {
+    SENSITIVE_KEYS.addAll(
+        Arrays.asList(
+            "token",
+            "assertion",
+            "access_token",
+            "client_secret",
+            "refresh_token",
+            "signedBlob",
+            "authorization"));
+  }
+
+  static void logRequest(HttpRequest request, LoggerProvider loggerProvider, String message) {
+    try {
+      Logger logger = loggerProvider.getLogger();
+      if (logger.isInfoEnabled()) {
+        Map<String, Object> loggingDataMap = new HashMap<>();
+        loggingDataMap.put("request.method", request.getRequestMethod());
+        loggingDataMap.put("request.url", request.getUrl().toString());
+
+        Map<String, Object> headers = new HashMap<>();
+        request
+            .getHeaders()
+            .forEach(
+                (key, val) -> {
+                  if (SENSITIVE_KEYS.contains(key)) {
+                    String hashedVal = calculateSHA256Hash(String.valueOf(val));
+                    headers.put(key, hashedVal);
+                  } else {
+                    headers.put(key, val);
+                  }
+                });
+        loggingDataMap.put("request.headers", gson.toJson(headers));
+
+        if (request.getContent() != null && logger.isDebugEnabled()) {
+          // are payload always GenericData? If so, can parse and store in json
+          if (request.getContent() instanceof UrlEncodedContent) {
+            // this is parsed to GenericData because that is how it is constructed.
+            GenericData data = (GenericData) ((UrlEncodedContent) request.getContent()).getData();
+            Map<String, Object> contextMap = parseGenericData(data);
+            loggingDataMap.put("request.payload", gson.toJson(contextMap));
+          } else if (request.getContent() instanceof JsonHttpContent) {
+            String jsonData = gson.toJson(((JsonHttpContent) request.getContent()).getData());
+            loggingDataMap.put("request.payload", jsonData);
+          }
+
+          Slf4jUtils.log(logger, org.slf4j.event.Level.DEBUG, loggingDataMap, message);
+        } else {
+
+          Slf4jUtils.log(logger, org.slf4j.event.Level.INFO, loggingDataMap, message);
+        }
+      }
+    } catch (Exception e) {
+      // let logging fail silently
+    }
+  }
+
+  static void logResponse(HttpResponse response, LoggerProvider loggerProvider, String message) {
+    try {
+      Logger logger = loggerProvider.getLogger();
+      if (logger.isInfoEnabled()) {
+        Map<String, Object> responseLogDataMap = new HashMap<>();
+        responseLogDataMap.put("response.status", String.valueOf(response.getStatusCode()));
+        responseLogDataMap.put("response.status.message", response.getStatusMessage());
```

Comment:
qq, the response message is nested as a field of status? Or should this be `request.message`?

Reply:
This is the status message, not a response message:
https://github.com/googleapis/google-http-java-client/blob/3ee8bb6380ee22a8a6eb6e5048c603fbdb9ee320/google-http-client/src/main/java/com/google/api/client/http/HttpResponse.java#L75-L76